app/assets/src/api/index.js

```diff
@@ -248,10 +248,78 @@ const createProject = params =>
   });
 
 // See https://czi.quip.com/bKDnAITc6CbE/How-to-start-instrumenting-analytics-2019-03-06
+// See also documentation for withAnalytics below.
 const logAnalyticsEvent = (eventName, eventData = {}) => {
   // Wrapper around Segment analytics so we can add things later
-  // eventData should have keys in snake_case for the database
-  if (window.analytics) window.analytics.track(eventName, eventData);
+  if (window.analytics) {
+    // Include high value user groups in event properties to avoid JOINs downstream.
+    if (window.analytics.user) {
+      const traits = window.analytics.user().traits();
+      eventData = {
+        // see traits_for_segment
+        admin: traits.admin,
+        biohub_user: traits.biohub_user,
+        czi_user: traits.biohub_user,
+        demo_user: traits.demo_user,
+        has_samples: traits.has_samples,
+        // caller can override above traits if they know what they are doing
+        ...eventData
+      };
+    }
+
+    // Evaluate lazy values, giving some time for React to set state
+    // See https://reactjs.org/docs/react-component.html#setstate
+    let shouldWait = false;
+    for (var k in eventData) {
+      if (typeof eventData[k] === "function") {
+        shouldWait = true;
+        setTimeout(() => (eventData[k] = eventData[k]()), 1);
```

Comment:
@tfrcarvalho this is a solution to a problem I ran into. For a call such as 

```
withAnalytics(
                    this.handleEmailChange,
                    "CreateUser_email_changed",
                    {
                      email: this.state.email
                    }
                  )
```

The `email` value is before being updated by `handleEmailChange` which is not what an analyst would want. Wrapping in function to make the value lazy `email: () => this.state.email` is still pretty clean.

But there is a second problem. React setState is async, so in the common case the above wouldn't work. Adding a `setTimeout` fixes the problem, at the cost of some indeterminism. We can't be 100% sure when React will be done. 

What do you think? If it is all too complicated, I'm happy to throw out the lazy evaluation and avoid setting this type of event-specific dynamic data until we actually have a need for it. Logging the event names and static data is 95% of the value right now.

Reply:
Yeah, I would avoid using state whenever possible.
IIUC, this particular example can be changed to `email: (event) => event.target.value`. I would actually whenever possible avoid reading from state, for the reasons you mention before and most create logs should come from event data. We are logging the action, not the processing.

Do you have a compelling example that cannot be refactored?

Reply:
No, I do not. I'm going to revert. Thanks!